vip-helpers/vip-newrelic.php

```diff
@@ -0,0 +1,134 @@
+	}
+}
+
+/**
+ * Name wp-cli correctly in New Relic and do not count it as part of the Apdex score
+ *
+ * We don't want to count ongoing WP-CLI requests as part of the apdex because it is not a user facing function and if a WP-CLI request is slow it doesn't imply that the site's performance is impacted. Without removing these WP-CLI requests from the apdex calculation it could flag the site as having performance problems, which would cause false positives in the monitoring.
+ */
+function wpcom_vip_wpcli_for_newrelic() {
+	if ( defined( 'WP_CLI' )
+	     && WP_CLI
+	     && ! wp_doing_cron()  // Cron is going to be run via WP-CLI in the near term. We want to keep Cron and WP-CLI separated for better monitoring so we're not going to flag WP_CLI requests that are actually cron requests as WP-CLI.
+	     && function_exists( 'newrelic_ignore_apdex' )
+	     && function_exists( 'newrelic_add_custom_parameter' )
+	     && function_exists( 'newrelic_name_transaction' ) ) {
+
+		newrelic_name_transaction( 'wp-cli' );
+
+		$wp_cli_arguments = \WP_CLI::get_runner()->arguments;
+		if ( empty( $wp_cli_arguments ) ) {
+			// Not much to do at this point
+			return;
+		}
+		if ( ! is_array( $wp_cli_arguments ) ) {
+			$wp_cli_arguments = [ $wp_cli_arguments ];
+		}
+		array_unshift( $wp_cli_arguments, 'wp' );
+		$cmd = implode( ' ', $wp_cli_arguments );
+		// e.g. `wp option get siteurl`
+		newrelic_add_custom_parameter( 'wp-cli-cmd', $cmd );
+
+		$assoc_args = \WP_CLI::get_runner()->assoc_args;
+		$config = \WP_CLI::get_runner()->config;
+		// Empty values clutter the result
+		foreach ( $config as $config_arg => $config_value ) {
+			if ( empty( $config_value ) ) {
+				unset( $config[$config_arg] );
+			}
+		}
+		$all_config = array_merge( $assoc_args, $config );
+		// Key sorting normalises the resultant string
+		ksort( $all_config );
+		// e.g. `allow-root=1&color=auto&format=json&skip-plugins=1` which is
+		// equivalent to `--allow-root --color=auto --format=json --skip-plugins`
+		// We have this format because building a query string is easy and understandable
+		// (and a hack) whereas rebuilding a WP CLI style command parameters is harder
+		newrelic_add_custom_parameter( 'wp-cli-cmd-args', urldecode(http_build_query( $all_config ) ) );
+
+		newrelic_ignore_apdex();
+	}
+}
+
+/**
+ * Name wp-api requests correctly in New Relic
+ *
+ * By default wp-api requests are tagged under index.php
+ * We'd want to have them tagged with the proper rest route used.
+ * While we are using the rest_dispatch_request filter, we're using it as an action without modifying the results.
+ */
+function wpcom_vip_rest_routes_for_newrelic( $dispatch_results, $request, $route, $handler ) {
+	if ( function_exists( 'newrelic_add_custom_parameter' )
+	     && function_exists( 'newrelic_name_transaction' ) ) {
+		newrelic_name_transaction( $route );
+		newrelic_add_custom_parameter( 'wp-api', 'true' );
+		newrelic_add_custom_parameter( 'wp-api-route', $route );
```

Comment:
Do we care that Cron Control will turn up here too? We only use the endpoints for monitoring at this point, so it won't skew things as much as including them with WP-CLI will.

Reply:
I don't think we care if it shows up or not. We can change it if needed.